docs/design/provisioning.md

```diff
@@ -0,0 +1,143 @@
+* Many ConfigMaps with "storage-class" as a key in the Data.  Difficult to validate unique storage names.
+* Name of ConfigMap as name of storage class. Naming validation is enforced by API server.  Requires PVC.Spec.StorageClass (or similar) with 1:1 mapping of claim to provisioner.
+
+
+
+### Plugin Parameters
+
+The example above contains `"params": "${OPAQUE-JSON-MAP}"`, which is a string representing a JSON map of parameters.   Each provisionable volume plugin will document and publish the parameters it accepts.
+
+Plugin params can vary by whatever attributes are exposed by a storage provider.  An AWS EBS volume, for example, can be a general purpose SSD (gp2), provisioned IOPs (io1) for speed and throughput, and magnetic (standard) for low cost.  Additionally, EBS volumes can optionally be encrypted.  Configuring each volume type with optional encryption would require 6 distinct provisioners.  Administrators can mix and match any attributes exposed by the provisioning plugin to create distinct storage classes.
+
+
+## Controller behavior
+
+`PersistentVolumeBinderController` continues to watch claims and attempts to bind them to volumes, creating them as necessary/possible.
+
+The binder attempts to fulfill a claim in the following order:
+
+* If `pvc.Spec.PersistentVolumeSelector` is non-nil and matches a provisioner's labels, create a PV using the provisioner
+* If `pvc.Spec.PersistentVolumeSelector` is non-nil but matches no provisioner, attempt to find an available PVs with a matching label set.
+* If `pvc.Spec.PersistentVolumeSelector` is nil, attempt to match an available PV by AccessModes and Capacity (current behavior).
+
+
+### Plugins
+
+Provisioning is implemented as plugins.  A number of standard plugins will be maintained within the Kubernetes codebase, but administrators may require implementations unique to their needs.  Future enhancements can include a DriverProvisioner to allow admins to create volumes that do not exist as plugins in the source tree.  The recent addition of FlexVolume is the prototype for this model.  That driver interface will improve over time and can eventually be made to provision volumes.
+
+The following plugins will be provided by Kubernetes:
+
+1.  EBSProvisioner -- creates a PV specifically for a claim and then creates the EBS volume in AWS.  Can be configured via parameters to create SSD, Provisioned IOPS, and Magnetic disks (EBS's 3 volume types).
+2.  GCEProvisioner -- creates a PV specifically for a claim and then creates the GCEPersistentDisk in GCE.  Can be configured via parameters to create pd-standard and pd-ssd disks.
+3.  CinderProvisioner -- creates a PV specifically for a claim and then creates a Cinder volume in OpenStack.  Can be configured via parameters to create multiple types (TODO: fill in exactly which types are supported)
+4. CephProvisioner -- creates a PV specifically for a claim and then creates the CephVolume in the infrastructure.
+
+
+# Tasks
```

Comment:
I think we need to add a section about mounting. Provisioners are things that dynamically create storage assets. Volume Plugins are things that we've historically referred to as something that mounts storage assets to containers within pods. If someone has a volume plugin that has no provisioner for it yet (such as the GlusterFS or NFS Volume Plugin) how does that person create a provisioner and tie the existing volume plugin to that provisioner so that when a claim is submitted that invokes the provisioner and a PV is created and bound to the claim, that when the claim is later used in a Pod, it knows which volume plugin to invoke for that type of PV. None of how this works is clear in this proposal.